lua/nvim-treesitter-textobjects/select.lua

```diff
@@ -1,11 +1,12 @@
 local api = vim.api
 local global_config = require "nvim-treesitter-textobjects.config"
 local shared = require "nvim-treesitter-textobjects.shared"
+local Range = require "vim.treesitter._range"
```

Comment:
This is a private API and so should not be used by plugins. @lewis6991

Reply:
I'm a bit confused about that. The [`Range4`](https://github.com/neovim/neovim/blob/01a56a056cd5da89350583ef9fbc97fa29dcd8ef/runtime/lua/vim/treesitter/_range.lua#L9) class defined in the same file (it should probably also be private) is used everywhere in this repository, but some of its related API functions were marked as private some was not, e.g. [`get_offset`](https://github.com/neovim/neovim/blob/01a56a056cd5da89350583ef9fbc97fa29dcd8ef/runtime/lua/vim/treesitter/_range.lua#L156).

However, these `unpack4()` functions just to avoid diagnostic warnings, I could just copy the code... I still recommend exposing them directly, they are just a collection of util functions.

Reply:
Private just means they don't have the same API guarantees in regard to things like deprecation strategy, documentation etc. I think it's fine to use them as long as those risks are accepted and understood. The whole of treesitter has less guarantees anyway so it's probably not as bad as it seems.

Also, I got a PR merged to LuaLS to improve the return type of unpack, not sure if that helps here.

Reply:
They are marked private (by Neovim) for a reason: the API is not considered final and may change at any time.

And I don't see any other private (prefixed with underscore) API currently (on  `main`)?

In any case, just making LuaLS shut up (which is a moving target) is not worth the risk of using unstable APIs.

Reply:
Moving _the whole_ plugin to use the upstream range API everywhere instead of custom classes inherited from nvim-treesitter.utils would be a different discussion, but one that should be had separately from type annotation fixes.

Reply:
> `Range4` is just a name local to this plugin; it does not refer to the upstream API.

This plugin itself did not define `Range4`, it should be `Range4` defined in `vim.treesitter`, and this is where Lua LS jumped to when called `textdocument/typeDefinition`.

Reply:
Yes, as I said, it's just a name (that happens to coincide with a "proper" class in core). There's a lot of refactors to replace custom nvim-treesitter code with newer Nvim APIs left on the table, and such PRs are very welcome. (They should just be done properly and separately, not partially as part of a type fix.)